Source/JavaScriptCore/jit/AssemblyHelpers.h

```diff
@@ -1518,26 +1518,42 @@ class AssemblyHelpers : public MacroAssembler {
         done.link(this);
     }
 
-    void branchConvertDoubleToInt52(FPRegisterID srcFPR, RegisterID destGPR, JumpList& failureCases, RegisterID scratch1GPR, FPRegisterID scratch2FPR)
+    void branchConvertDoubleToInt52(FPRegisterID srcFPR, RegisterID destGPR, JumpList& failureCases, RegisterID scratch1GPR, FPRegisterID scratch2FPR, bool canIgnoreNegativeZero)
     {
         JumpList doneCases;
 
         truncateDoubleToInt64(srcFPR, destGPR);
-        convertInt64ToDouble(destGPR, scratch2FPR);
+
+        bool convertedBack = false;
+#if CPU(ARM64)
+        if (supportsRoundFloatToIntegerFloat()) {
+            convertedBack = true;
+            roundTowardZeroInt64Double(srcFPR, scratch2FPR);
+        }
+#endif
+        if (!convertedBack)
+            convertInt64ToDouble(destGPR, scratch2FPR);
+
         failureCases.append(branchDouble(DoubleNotEqualOrUnordered, srcFPR, scratch2FPR));
-        auto isZero = branchTest64(Zero, destGPR);
+
+        Jump isZero;
+        if (!canIgnoreNegativeZero)
+            isZero = branchTest64(Zero, destGPR);
+
         // This moves the checking range (fail if N >= (1 << (52 - 1)) or N < -(1 << (52 - 1))) by subtracting a value.
         // So, valid value region starts with -1 and lower. In unsigned form, which means,
         // 0xffffffffffffffff to 0xfff0000000000000. So, by shifting 52, we can extract 0xfff part, and we can check whether it is below than that (<= 4094).
         move(TrustedImm64(0xfff8000000000000ULL), scratch1GPR);
         add64(destGPR, scratch1GPR);
         urshift64(TrustedImm32(52), scratch1GPR);
         failureCases.append(branch64(BelowOrEqual, scratch1GPR, TrustedImm32(4094)));
```

Comment:
Why do we check `4094` instead of `4095` (`2^12 - 1`)? 

For example, if we check whether `int16` is in `int8` range (from `-7` to `8`), then we do similar things like above
```
7 - 8 = -1 // 1111 1111
-8 - 8 = -16 // 1111 0000
```
then unsigned right shift by 4 bits and both should have
```
0000 1111
```
which means if the result is `<= 15` (`2^4 - 1`) then it's under `int8` range. Similar to check whether `int64` under `int52` range. We should check `<= 4095` (`2^12 - 1`), right?

Reply:
The range is not -7 to 8. If the number is signed, negative one always has larger range, so the range is -8 to 7.

Reply:
Oops, corrected my previous comment. The calculation is correct right?

Reply:
I think `4094` here is right code.
The range in Int52 in hex representation is,

```
negative: 0xfff8000000000000 - 0xffffffffffffffff
positive: 0x0000000000000000 - 0x0007ffffffffffff
```
By adding `0xfff8000000000000`, how they get changed is
```
0xfff8000000000000 => 0xfff0000000000000
0xffffffffffffffff => 0xfff7ffffffffffff
0x0000000000000000 => 0xfff8000000000000
0x0007ffffffffffff => 0xffffffffffffffff
```

This means that the effective range is now shifted to
```
0xfff0000000000000 - 0xffffffffffffffff
```
And when extracting the above bits, then it should be `0xfff0000000000000 >> 52` = `0xfff`
So, if the bit is `0xfff`, then it is right. If it is less than 0xfff, then that's wrong. So `<= 4094` is correct.

Reply:
I see `<= 4094` is for the false case, and `> 4094` is the valid case which `4095` is included. Thanks.